lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py

```diff
@@ -0,0 +1,533 @@
+            empirical cumulative distribution function i.e. probabilities
+            below threshold.
+
+        Returns
+        -------
+        threshold_points : Numpy array
+            Array of threshold values padded with the lower and upper bound.
+        probabilities_for_cdf : Numpy array
+            Array containing the probabilities padded with 0 and 1 at each end.
+        bounds_pairing : Tuple
+            Lower and upper bound to be used as the ends of the
+            empirical cumulative distribution function.
+
+        """
+        lower_bound, upper_bound = bounds_pairing
+        threshold_points = np.insert(threshold_points, 0, lower_bound)
+        threshold_points = np.append(threshold_points, upper_bound)
+        zeroes_array = np.zeros((probabilities_for_cdf.shape[0], 1))
+        ones_array = np.ones((probabilities_for_cdf.shape[0], 1))
+        probabilities_for_cdf = np.concatenate(
+            (zeroes_array, probabilities_for_cdf, ones_array), axis=1)
+        return threshold_points, probabilities_for_cdf
+
+    def _probabilities_to_percentiles(
+            self, forecast_probabilities, percentiles, bounds_pairing):
+        """
+        Conversion of probabilities to percentiles through the construction
+        of an empirical cumulative distribution function. This is effectively
+        constructed by linear interpolation from the probabilities associated
+        with each threshold to a set of percentiles.
+
+        Parameters
+        ----------
+        forecast_probabilities : Iris cube
+            Cube with a probability_above_threshold coordinate.
+        percentiles : Numpy array
+            Array of percentiles, at which the corresponding values will be
+            calculated.
+        bounds_pairing : Tuple
+            Lower and upper bound to be used as the ends of the
+            empirical cumulative distribution function.
+
+        Returns
+        -------
+        percentile_cube : Iris cube
+            Cube with probabilities at the required percentiles.
+
+        """
+        threshold_points = (
+            forecast_probabilities.coord("probability_above_threshold").points)
+
+        prob_slices = convert_cube_data_to_2d(
+            forecast_probabilities, coord="probability_above_threshold")
+
+        # Invert probabilities
+        probabilities_for_cdf = 1 - prob_slices
+
+        threshold_points, probabilities_for_cdf = (
+            self._add_bounds_to_thresholds_and_probabilities(
+                threshold_points, probabilities_for_cdf, bounds_pairing))
+
+        forecast_at_percentiles = (
+            np.empty((probabilities_for_cdf.shape[0], len(percentiles))))
+        for index in range(probabilities_for_cdf.shape[0]):
+            forecast_at_percentiles[index, :] = np.interp(
+                percentiles, probabilities_for_cdf[index, :],
+                threshold_points)
+
+        # Reshape forecast_at_percentiles, so the percentiles dimension is
```

Comment:
At the moment, this hasn't been factored out, as it doesn't particularly make a clean utility, although this may be reconsidered.